File: 11.Con-Water/cwmw.js
/*global siyu*/
/**
 * time limit exceeded!!
 */
var maxAreaOne = function (height) {
    var arr = height.slice();
    var length = arr.length;
    var max = 0;

    var getContain = function (left, right) {
        var height = right - left;
        var side = Math.min(arr[left], arr[right]);
        return side * height;
    };

    for (var i = 0; i < length - 1; i++) {
        for (var j = i + 1; j < length; j++) {
            var contain = getContain(i, j);
            if (contain > max) {
                max = contain;
            }
        }
    }

    return max;
};

/**
 * accepted
*/
var maxAreaTwo = function (height) {
    var l = 0;
    var r = height.length - 1;
    var max = 0;

    while (l < r) {
        var left = height[l];
        var right = height[r];
        var side = Math.min(left, right);
        var tall = r - l;
        if (left > right) {
            r = r - 1;
        } else {
            l = l + 1;
        }
        max = Math.max(side * tall, max);
    }

    return max;
};

console.log(maxAreaTwo([1, 1]));

File: 11.Con-Water/cwmw.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int maxArea(vector<int>& height) {
    int l = 0;
    int r = height.size() - 1;
    int water = 0;
    while (l < r) {
        int side = min(height[l], height[r]);
        water = max(water, side * (r - l));
        if (height[l] <= height[r]) {
            l = l + 1;
        } else {
            r = r - 1;
        }
    }
    return water;
}

int main() {
    int arr[] = {2, 1, 3};
    vector<int> a(arr, arr + 3);
    cout << maxArea(a) << endl;
    return 0;
}